dev-docs/patches.md

```diff
@@ -0,0 +1,95 @@
+# Working with patches
+
+## Patches in Nix
+
+There are two common ways to apply patches in Nix:
+
+- Use the `patches` attribute of a builder ([manual](https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-patches)):
+
+  ```nix
+  stdenv.mkDerivation {
+    # ...
+    patches = [
+        ./0001-first-change.patch
+        ./0002-another-change.patch
+    ]
+    # ...
+  }
+  ```
+  This is most commonly used and can be encountered nearly everywhere upstream in nixpkgs.
+
+- Use the `applyPatches` function ([noogle](https://noogle.dev/f/pkgs/applyPatches)) to patch an existing source and derive a new source:
+
+  ```nix
+  stdenv.mkDerivation {
+    # ...
+    src = applyPatches {
+        src = fetchFromGuiHub { /* ... */ }
+    patches = [
+        ./0001-first-change.patch
+        ./0002-another-change.patch
+    ]
+    # ...
+  }
+  ```
+
+  This is useful in situations where the patches sources should also be used by other consumers, as the
+  patched source can inherited, but `patches` aren't part of the output set.
+
+Patches can either be consumed from a local file or pulled from remote:
+
+```nix
+patches = [
+  (fetchpatch {
+    url = "https://github.com/...";
+    stripLen = 2; # Sometimes the patch strip level must be configured
+    hash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
+  })
+];
+```
+
+If there aren't any changes needed to an upstream patch, it's good practice to fetch the patch from remote. If the changes to upstream patches are needed or patches aren't available upstream, the
+*full patch set* should be vendored, so it can easily be worked with.
+
+## Patch set development flow
+
+For a smooth development experience, we only use `git format-patch` format for patches. This enables us
+to apply the changes to the source, work on it and sync back.
+
+Given a package with some existing patches in `pkgDir` (for example `packages/by-name/contrast`).
+
+Clone the source and checkout the `rev` currently used in the package `src` (the commit or tag):
+
+```sh
+git clone $url
+git checkout $rev
+```
+
+Apply the existing patch set:
+
+```sh
+git am -3 $pkgDir/*.patch
+```
+
+This will apply and commit each patch on top of `rev`.
+
+You can then place new commits on top or modify existing commits. Remember to keep the git history clean.
+
+When updating a package, you might need to rebase the current patch set.
+
+When done, recreate the patch set:
+
+```sh
+git format-patch -o $pkgDir $rev
+```
+
+# Patch documentation conventions
+
+Patches need thorough documentation. Each reference of a patch must have a comment.
```

Comment:
> Each reference of a patch must have a comment.

You mean a comment in the `.patch` file, after the `---`?

Reply:
I've added an example and some reasoning about it.